internal/logging/logging.go

```diff
@@ -0,0 +1,134 @@
+// Package logging implements the logger for the pack CLI.
+package logging
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/apex/log"
+
+	"github.com/buildpack/lifecycle/style"
+)
+
+const (
+	InfoLevel  = "info"
+	DebugLevel = "debug"
+
+	errorLevelText = "ERROR: "
+	warnLevelText  = "Warning: "
+
+	// time format the out logging uses
+	timeFmt = "2006/01/02 15:04:05.000000"
+)
+
+// handler implementation.
+type handler struct {
+	sync.Mutex
+	writer   io.Writer
+	wantTime bool
+	timer    func() time.Time
+}
+
+func formatLevel(ll log.Level) string {
+	switch ll {
+	case log.ErrorLevel:
+		return style.Error(errorLevelText)
+	case log.WarnLevel:
+		return style.Warn(warnLevelText)
+	}
+
+	return ""
+}
+
+// preserve behavior of other loggers
+func appendMissingLineFeed(msg string) string {
+	buff := []byte(msg)
+	if buff[len(buff)-1] != '\n' {
+		buff = append(buff, '\n')
+	}
+	return string(buff)
+}
+
+// HandleLog supports behavior that is unique to Pack CLI, namely toggling colors and timestamps.
+func (h *handler) HandleLog(e *log.Entry) error {
+	h.Lock()
+	defer h.Unlock()
+
+	// if we have a blank line we don't want padding or prefixes
+	if e.Message == "" {
+		_, _ = fmt.Fprintln(h.writer)
+		return nil
+	}
+
+	if h.wantTime {
+		ts := h.timer().Format(timeFmt)
+		_, _ = fmt.Fprint(h.writer, appendMissingLineFeed(fmt.Sprintf("%s %s%-25s", ts, formatLevel(e.Level), e.Message)))
+		return nil
+	}
+
+	_, _ = fmt.Fprint(h.writer, appendMissingLineFeed(fmt.Sprintf("%s%-25s", formatLevel(e.Level), e.Message)))
+
+	return nil
+}
+
+type logWithWriters struct {
+	log.Logger
+	out     io.Writer
+	errOut  io.Writer
+	handler *handler
+}
+
+func (lw *logWithWriters) Writer() io.Writer {
+	return lw.out
+}
+
+// DebugErrorWriter - returns stderr if log level is not set to quiet.
+func (lw *logWithWriters) DebugErrorWriter() io.Writer {
+	if lw.Level == log.DebugLevel {
+		return lw.errOut
+	}
+	return ioutil.Discard
+}
+
+// DebugWriter returns stdout if logging is not set to quiet.
+func (lw *logWithWriters) DebugWriter() io.Writer {
+	if lw.Level == log.DebugLevel {
+		return lw.out
+	}
+	return ioutil.Discard
+}
+
+func (lw *logWithWriters) WantTime(f bool) {
+	lw.handler.wantTime = f
+}
+
+func (lw *logWithWriters) WantLevel(f string) {
+	if f == InfoLevel {
+		lw.Level = log.InfoLevel
+	} else if f == DebugLevel {
+		lw.Level = log.DebugLevel
+	} else {
+		lw.Level = log.ErrorLevel
+	}
+}
+
+// NewLogWithWriters creates a logger to be used with pack CLI.
+func NewLogWithWriters() *logWithWriters {
```

Comment:
Returning an unexported struct can make life difficult for the consumer. They are unable to type the object they receive. Instead, return the struct through the interface if there's a standard interface for it (`log.Logger`). 

It's often an anti-pattern to define an interface and unexported struct in the same package. But not in this case, where the purpose is to a define a standard logging interface, and provide one possible implementation of it.

Reply:
yea, i was wondering about that, but I was cribbing from `pack`. i'll fix here, but wonder about pack too

Reply:
hmm, actually, I do need to invoke `WantLevel` outside this package. Should we just export `logWithWriters`?

Reply:
Could all external input be accomplished via the interface or via the constructor?

Reply:
E.g., you could use functional options in the constructor: 
https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Reply:
Yea, I can do that, but then I'm really starting to diverge from what we're doing in pack. Is that ok?

Reply:
I think it's a good pattern and we'll probably want to move pack towards it anyway.

Reply:
The problem I'm up against is that `NewLogWithWriters()` is called before we know the logging level. I can remove `cmd.Logger` and have each `main.go` create it's own instance once it gets the flag. Or is there another way?

Reply:
> Yea, I can do that, but then I'm really starting to diverge from what we're doing in pack. Is that ok?

... and here start the problems... We've made minor changes to pack logger (some more in review) since this PR opened. We really need to determine what the best way of minimizing maintenance.

Reply:
I've added `WantLevel(string)` to the `Logger` interface.